module-system/module-implementations/sov-attester-incentives/src/call.rs

```diff
@@ -0,0 +1,738 @@
+            let finality_period = self
+                .rollup_finality_period
+                .get(working_set)
+                .expect("Should be defined at genesis");
+
+            ensure!(
+                begin_unbond_height + finality_period < curr_height,
+                "Cannot unbond the attester: finality has not completed yet"
+            );
+
+            // Get the user's old balance.
+            if let Some(old_balance) = self.bonded_challengers.get(context.sender(), working_set) {
+                // Transfer the bond amount from the sender to the module's address.
+                // On failure, no state is changed
+                self.reward_sender(context, old_balance, working_set)?;
+
+                // Update our internal tracking of the total bonded amount for the sender.
+                self.bonded_challengers
+                    .remove(context.sender(), working_set);
+                self.unbonding_attesters
+                    .remove(context.sender(), working_set);
+
+                // Emit the unbonding event
+                working_set.add_event(
+                    "unbonded_challenger",
+                    &format!("amount_withdrawn: {old_balance:?}"),
+                );
+            }
+        }
+
+        Ok(CallResponse::default())
+    }
+
+    /// The bonding proof is now a proof that an attester was bonded during the last `finality_period` range.
+    /// The proof must refer to a valid state of the rollup. The initial root hash must represent a state between
+    /// the bonding proof one and the current state.
+    fn check_bonding_proof(
+        &self,
+        context: &C,
+        attestation: &Attestation<StorageProof<<C::Storage as Storage>::Proof>>,
+        working_set: &mut WorkingSet<C::Storage>,
+    ) -> Result<(), AttesterIncentiveErrors> {
+        let bonding_root = {
+            // If we cannot get the transition before the current one, it means that we are trying
+            // to get the genesis state root
+            if let Some(transition) = self
+                .chain_state
+                .historical_transitions
+                .get(&(attestation.proof_of_bond.transition_num - 1), working_set)
+            {
+                transition.post_state_root()
+            } else {
+                self.chain_state
+                    .genesis_hash
+                    .get(working_set)
+                    .expect("The genesis hash should be set at genesis")
+            }
+        };
+
+        // This proof checks that the attester was bonded at the given transition num
+        let (attester_key, bond_opt) = working_set
+            .backing()
+            .open_proof(bonding_root, attestation.proof_of_bond.proof.clone())
+            .map_err(|_err| AttesterIncentiveErrors::InvalidBondingProof)?;
+
+        // We have to check that the storage key is the same as the sender's
+        if attester_key
+            != StorageKey::new(
```

Comment:
The `StorageKey` is a low level concept and we should avoid using it  in the module code. Would it work if instead of `open_proof` we have `verify_proof` which does this check and the one below (for bond)? This way `StorageKey` and deserialisation logic won't leak to the module code.